source/nodejs/adaptivecards/src/markdown-formatter.ts

```diff
@@ -1,182 +1,186 @@
-		},
+            patterns: ['[]()'],
+            patternType: 'asymmetric',
+            groups: 2,
+        }
     ];
 
-    //combines default config and user config.
-    static setCustomMarkdownRegex(regexArray: any[]){
-        MarkdownFormatter.regexArray = regexArray;
-
-		// prefer user configs
-		for(var i = 0; i < MarkdownFormatter.MD_FORMATTER_CONFIG.length; i++){
-			for(var j =0; j < MarkdownFormatter.regexArray.length; j++){
-				if(MarkdownFormatter.MD_FORMATTER_CONFIG[i].type == MarkdownFormatter.regexArray[j].type){
-					MarkdownFormatter.MD_FORMATTER_CONFIG[i] = MarkdownFormatter.regexArray[j];
-					MarkdownFormatter.regexArray.splice(j, 1);
-					continue;					
-				}
-			}
-		}
-        MarkdownFormatter.regexArray = MarkdownFormatter.MD_FORMATTER_CONFIG.concat(MarkdownFormatter.regexArray);
-        
+    constructor() {
+        super();
+        this.configuration.forEach(config => {
+            this.initailize(config);
+        });
     }
-    
-
-	constructor(text:string) {
-        this.text = text;
-       
-		for (var i = 0; i < MarkdownFormatter.MD_FORMATTER_CONFIG.length; i++) {
-			let patternArray = MarkdownFormatter.MD_FORMATTER_CONFIG[i].pattern;
-			let patternType = MarkdownFormatter.MD_FORMATTER_CONFIG[i].patternType;
-            let groups = MarkdownFormatter.MD_FORMATTER_CONFIG[i].groups;
-            let pattern = patternArray[0];
-            
-            if(patternType === "custom"){                
-                pattern = patternArray[0];
+
+    /**
+     * To add custom markdown rules. This will update the default markdown rule is the same type is passed.
+     * 
+     * @param configurations 
+     */
+    public setCustomRules(configurations: IFormatterConfiguration[]) {
+        for (var j = 0; j < configurations.length; j++) {
+            let index = this.configuration.findIndex(config => config.type == configurations[j].type);
+            if (index >= 0) {
+                this.configuration[index] = configurations[j];
+                this.initailize(configurations[j]);
+                continue;
+            } else {
+                this.configuration.push(configurations[j]);
+            }
+        }
+    }
+
+    /**
+     * To add '\' before all the special characters in the regex.
+     * 
+     * @param pattern 
+     */
+    private encodeSpecialCharacters(pattern: string) {
+        return pattern.replace(MarkdownFormatter.SPECIAL_CHAR_REGEX, "\\$&")
+    }
+
+    private initailize(configuration: IFormatterConfiguration): void {
+
+        let patternArray = configuration.patterns,
+            patternType = configuration.patternType,
+            pattern = patternArray[0],
+            groups = configuration.groups;
+
+        if (patternType === "list") {
+            pattern = this.encodeSpecialCharacters(patternArray[0]) + '[\\s]+((.*?)[' + this.encodeSpecialCharacters(patternArray[1]) + "](?=" + this.encodeSpecialCharacters(patternArray[1]) + "[\\s]+)|(.*?)$)";
```

Comment:
That's better, but now `this.encodeSpecialCharacter(patternArray[0])` and `this.encodeSpecialCharacter(patternArray[1])` should really be put in variables given there are called 4 and 2 times respectively.